proxy/Dockerfile

```diff
@@ -1,24 +1,106 @@
 # Proxy build and runtime
 #
-# Builds a slim runtime image with the conduit-proxy binary.
+# When BUILD_UNOPTIMIZED is set and not empty, unoptimized rust artifacts are produced.
+# This reduces build time and produces binaries with debug symbols, at the expense of
+# runtime perforamnce.
+#
+# When SKIP_TESTS is set and not empty, tests are not run. Otherwise, tests are run
+# against either unoptimized or optimized proxy code, according to BUILD_UNOPTIMIZED.
+
+ARG RUST_IMAGE=rust:1.23.0
+
+ARG RUNTIME_IMAGE=gcr.io/runconduit/base:2017-10-30.01
+
+## Fetch external dependencies.
+#
+# Mocks out all local code and fetch external dependencies to ensure that external sources
+# are cached.
+FROM $RUST_IMAGE as fetch
+WORKDIR /usr/src/conduit
+RUN for d in proxy proxy/controller-grpc proxy/convert proxy/futures-mpsc-lossy proxy/router ; \
+    do mkdir -p "${d}/src" && touch "${d}/src/lib.rs" ; \
+    done
+COPY Cargo.toml Cargo.lock ./
+COPY proxy/Cargo.toml                       proxy/Cargo.toml
+COPY proxy/controller-grpc/Cargo.toml       proxy/controller-grpc/Cargo.toml
+COPY proxy/convert/Cargo.toml               proxy/convert/Cargo.toml
+COPY proxy/futures-mpsc-lossy/Cargo.toml    proxy/futures-mpsc-lossy/Cargo.toml
+COPY proxy/router/Cargo.toml                proxy/router/Cargo.toml
+RUN cargo fetch --locked
+
+## Build libraries.
+#
+# Mocks proxy sources and gRPC bindings to ensure that dependencies are cached.
+FROM $RUST_IMAGE as libs
```

Comment:
Please add a comment about why this is a separate named stage, instead of just being the end of the previous stage.

It seems like everything should work just fine if `fetch` and `libs` were combined into a single stage. Am I overlooking something?